### ISSUE TITLE
[8.0][web_widget_x2many_2d_matrix] Cannot read property 'id' of undefined

### ISSUE THREAD

**Issue by DayssamAgora** [OCA/web#321]
Hi everyone,

I got this error when i want to re show the matrix

Uncaught TypeError: Cannot read property 'id' of undefined
http://localhost:3000/web_widget_x2many_2d_matrix/static/src/js/web_widget_x2many_2d_matrix.js:215

I debuged the javascript code

the error was in this function

`// return row id of a coordinate
        get_xy_id: function(x, y)
        {
            return this.by_x_axis[x][y]['id'];
        },`

when i logged the x & y values

It gives me the value of x is "undefined"

x=1 y=1
x=2 y=1
x=3 y=1
x=4 y=1
x=5 y=1
x=6 y=1
x=7 y=1
x=8 y=1
x=9 y=1
x=10 y=1
x=11 y=1
x=undefined y=1

Normally i have only 11 rows
and it iterates only the first column

**Comment by hbrunn** [comment]
This widget is quite picky about the data. What code do you use to populate the field and how does the internal data structure look?

**Comment by DayssamAgora** [comment]
![matrix2d](https://cloud.githubusercontent.com/assets/16352291/13634507/e226c882-e5ed-11e5-8bc6-e76719f0d12e.png) 

The python code :

```
@api.model
def default_get(self,fields):
    defaults = super(Saisie,self).default_get(fields)
    default = super(Saisie,self).default_get(['projet_id'])
    if default:
        projet=self.env['projet.projet'].search([('id','=',default['projet_id'])])
        model=projet.model_saisie
        marques=model.marque_ids
        parametres=model.parametre_ids
        ids= [rec.id for rec in marques]
        marque_names_list = [rec.name for rec in self.env['saisie.marque'].browse(ids) ]
        ids2= [rec.id for rec in parametres]
        parametre_names_list = [rec.name for rec in self.env['saisie.parametre'].browse(ids2)]        
        print "p list", parametre_names_list
        inputdata = [
        parametre_names_list,
        [0],
        marque_names_list,
        [0]
        ]
        result = list(itertools.product(*inputdata))
        print(result)
        defaults['saisie_ligne_id'] = [
        (0, 0, {
            'value': 0,
            'parametre_id': p,
            'marque_id': e,
        })
        for e in ids
        for p in ids2
        ] 
    return defaults
```

The XML CODE : 

<field name="saisie_ligne_id" widget="x2many_2d_matrix"
                            field_x_axis="parametre_id" field_y_axis="marque_id" field_value="value"
                            attrs="{'readonly': [('state', '=', 'valide')]}" />

**Comment by DayssamAgora** [comment]
@hbrunn :

I will make a video showing you the scenario which produce the error

**Comment by hbrunn** [comment]
that will be helpful. From your code it looks very good, I don't see anything that might be causing the problem

**Comment by DayssamAgora** [comment]
@hbrunn 
https://youtu.be/s_W5yhPilLk

The error was in this function

// return row id of a coordinate get_xy_id: function(x, y) { return this.by_x_axis[x][y]['id']; },

when i logged the x & y values

It gives me the value of x is "undefined"

x=1 y=1
x=2 y=1
x=3 y=1
x=4 y=1
x=5 y=1
x=6 y=1
x=7 y=1
x=8 y=1
x=9 y=1
x=10 y=1
x=11 y=1
x=undefined y=1

**Comment by hbrunn** [comment]
interesting. So something seems to go wrong if the widget is build repeatedly. Do you see any patterns here, like the 3rd time you open this, the error occurs? Any patterns in the values filled in? And does a reload fix the problem temporarily?

**Comment by DayssamAgora** [comment]
Always the error occurs. but when i save the form there is no problem,

**Comment by DayssamAgora** [comment]
@hbrunn 

The error occurs in the changed field.

when i change x1y1 it stops @ x1y1

when i change x3y1 it  stops @ x3y1

**Comment by DayssamAgora** [comment]
@hbrunn 

Any idea about sorting the x fields or y fields (not sorted by id by default) ?

**Comment by hbrunn** [comment]
hmmm, I'm running out of ideas here. I think I need to debug what's going on here. Could you please provide a minimal example module that exposes this behavior when being installed in an empty database with demo data? Then we should be able to fix the problem.

**Comment by DayssamAgora** [comment]
@hbrunn 
Can i send it to you in private ?

**Comment by hbrunn** [comment]
yes, makes sense, you should find my mail address on the github profile or the oca site

**Comment by DayssamAgora** [comment]
@hbrunn 
Its done

**Comment by hbrunn** [comment]
@DayssamAgora this should be fixed in #327. Next time somebody asks you for a minimal example that produces the error, please actually provide a minimal example. I had to dive through a lot of unnecessary code, and you're lucky I was curious about that problem, otherwise I'd have put this back to you

**Comment by mhabou** [comment]
@hbrunn :+1: 

Many thanks ,

I will post a minimal example that produce the error.

**Comment by hbrunn** [comment]
now it's solved, so not necessary now. That was just a tip for the future

**Comment by hbrunn** [comment]
but please review my PR and if it solves your problem, give a thumbs up there. This issue will be closed automatically when the code is merged

**Comment by DayssamAgora** [comment]
it's done :+1:

**Comment by hbrunn** [comment]
fixed by merging #327 a while ago
